### PR TITLE
Add new modal options to require confirmation before closing

### PR DESCRIPTION
**Pull request by davwheat**
> For clarity, this PR was originally for adding attrs which disabled modal dismissing via backdrop clicks and Esc keypresses, but was changed after discussion.

I'm aware that Bootstrap Modals is set to get replaced by Micromodal (or another library) in the near future, but implementing this functionality now should (hopefully) mean it gets carried across to the new library. 🤞 

**Changes proposed in this pull request:**
This provides more customisability for modals, allowing extensions (and core) to determine if confirmation should be required before closing the modal.

The current field `isDismissible` was previously incorrectly labelled as only determining if an X button is visible, not stating that it also controlled if a backdrop click or Esc key press would dismiss the modal. (To be fair, I can't think of any situation where you want to remove a close button and keep Esc and backdrop dismissal.)

This PR adds two new static fields: `requireImplicitCloseConfirmation` and `requireExplicitCloseConfirmation`. This allows extensions to set whether confirmation should be needed for implicit actions, such as backdrop clicks and Esc keypresses.

I've been in the middle of completing several modals with forms in some extensions where I begin to click the button, then decide to edit one of the form entries, so drag my cursor off the submit button onto the backdrop, which then closes the modal. I then either have to re-complete the entire modal again, or just don't bother.

Another situation where backdrop dismissals frustrate me is if I'm copying text from the modal. I tend to click and drag off the end of the component (bottom or sides) to quickly copy all the text I need quickly. Releasing the click causes the modal to dismiss.

While extensions should probably find a way around this, or Bootstrap itself should be patched, in the long run, more customisability is better (in my mind), so exposing these options to extensions is a good thing!

---

This also addresses an issue with Esc-based closure. BS Modals currently only detect Esc presses when performed within their content, which requires that they have focused content. Modals without any autofocused content (no inputs, selects, or textareas) cannot be closed via Esc by default.

This PR disables the default detection techniques used by BS Modal and instead implements custom handlers to match behaviour between the confirmation-based and standard closures, and to fix this issue.

https://user-images.githubusercontent.com/7406822/111080808-0d9bc400-84f8-11eb-8c0e-d0e7fac5e2cf.mp4

**Reviewers should focus on:**
Where should `requireImplicitCloseConfirmation` be overridden to `true`? I've thought of the tag selection modal off the top of my head, but please suggest any other locations.

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.

### CODE REVIEW

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
Looks like some dist changes were accidentially included, could you remove those before we get into review?

File: js/src/common/components/ModalManager.js [hunk 1]
```diff
@@ -33,7 +33,11 @@ export default class ModalManager extends Component {
   }
 
   animateShow(readyCallback) {
-    const dismissible = !!this.attrs.state.modal.componentClass.isDismissible;
+    const isDismissible = !!this.attrs.state.modal.componentClass.isDismissible;
+
+    // If the modal isn't dismissible, set these options to false.
+    const backdropDismissible = isDismissible ? !!this.attrs.state.modal.componentClass.doesBackdropClickDismiss : false;
```
Comment (askvortsov1) [inline js/src/common/components/ModalManager.js:39]: Couldn't these be `isDismissible && !!theAttr`?

File: js/src/common/components/Modal.js [hunk 2]
```diff
@@ -10,10 +10,28 @@ import Button from './Button';
...
+  /**
+   * Determine whether a click on the modal backdrop should dismiss the modal.
+   *
+   * Clicking and dragging from certain elements onto the backdrop can sometimes
+   * inadvertantly trigger this.
+   *
+   * It's recommended to disable this for modals which contain forms that may
+   * take a user a longer amount of time to complete to prevent frustration.
+   */
+  static doesBackdropClickDismiss = false;
```
Comment (askvortsov1) [inline js/src/common/components/Modal.js:28]: How do we feel about this wording vs `dismissOnBackdropClick`?

File: js/src/common/components/ModalManager.js [hunk 3]
```diff
@@ -46,10 +50,23 @@ export default class ModalManager extends Component {
     this.$()
       .one('shown.bs.modal', readyCallback)
       .modal({
-        backdrop: dismissible || 'static',
-        keyboard: dismissible,
+        // 'static' means that a backdrop click doesn't dismiss the modal
+        backdrop: backdropDismissible ? true : 'static',
```
Comment (askvortsov1) [inline js/src/common/components/ModalManager.js:54]: Why is this change necessary?

File: js/src/common/components/ModalManager.js [hunk 4]
```diff
@@ -46,10 +50,23 @@ export default class ModalManager extends Component {
...
       .modal('show');
+
+    // Disabling backdrop dismissal also disables keyboard dismissibility
+    // but we want to be able to provide key-based dismissal while having
+    // backdrop-based disabled. To do this, we set up our own event
+    // handler to hide the modal if we press Escape.
+    if (!backdropDismissible && keyboardDismissible) {
+      this.$().on('keypress', function (e) {
+        if (e.key === 'Escape') {
+          this.animateHide();
```
Comment (askvortsov1) [inline js/src/common/components/ModalManager.js:66]: Not a change request but a comment for reviewers: we'll need to test that dismissing this way doesn't result in weird UI glitches.

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

**Comment by davwheat** [comment]
Dist changes totally my fault... I think I got too greedy with `git commit -a`! On that now.

File: js/src/common/components/ModalManager.js (hunk 1 above)
Comment (davwheat) [reply js/src/common/components/ModalManager.js:39]: Yes, it could and it definitely should!

File: js/src/common/components/Modal.js (hunk 2 above)
Comment (davwheat) [reply js/src/common/components/Modal.js:28]: That name sounds way better in my opinion. I think we should go for it.

File: js/src/common/components/ModalManager.js (hunk 3 above)
Comment (davwheat) [reply js/src/common/components/ModalManager.js:54]: Oops, it isn't. I think this slipped through the net while I was trying other approaches.

I'll revert back to the `||` method, but keep the comment as it makes the reasoning behind `"static"` clearer.

File: js/src/common/components/ModalManager.js (hunk 1 above)
Comment (davwheat) [reply js/src/common/components/ModalManager.js:39]: I just realised that when `isDismissible` is true and backdrop dismissal is false, the modal will still close on backdrop click if using the `||` method as `isDismissible` will override the custom properties.

My intention was to only override if the dialog was not dismissible, otherwise we're in the same situation as before this PR.

Comment (askvortsov1) [reply js/src/common/components/ModalManager.js]: But wouldn't it work with `&&`?

Comment (davwheat) [reply js/src/common/components/ModalManager.js]: Yes... if only I didn't misread what you wrote 😅

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
Had a chance to test this out. I get where this is coming from with a desire to prevent accidential closings, but not being able to click outside of the modal to close it is problematic IMO. It definitely shouldn't be the default behavior, but I'm not even sure whether we want to make it an option, since it might be confusing to users.

**Comment by davwheat** [comment]
Maybe instead of fully disabling closing via backdrop click, it should enable a prompt, similar to discarding a post draft?

**Comment by askvortsov1** [comment]
> Maybe instead of fully disabling closing via backdrop click, it should enable a prompt, similar to discarding a post draft?

I think a "confirm modal close" system would be great!

**Comment by davwheat** [comment]
Whoops... accidentally closed when I was cleaning up my fork 🙈

**Comment by askvortsov1** [comment]
How would this look if we proceeded with https://github.com/flarum/core/issues/2345#issuecomment-798904086? Would it still be necessary / maybe it might make sense to implement it after the modal rewrite, if we can get it into the release?

**Comment by davwheat** [comment]
It's an interesting point. My initial idea was to implement this first just so that, when I get round to implementing native `<dialog>` elements, that this gets carried over to maintain the same API, but if you'd rather put it all in one, that's fine with me.

**Comment by askvortsov1** [comment]
Separate PRs is preferably, but you raise a good point about API: even if the new approach fixes closing if clicking on the modal and then dragging mouse outside, the API would be consistent.

The biggest reason I'm wary about this is that it feels `added  on`. What if instead we had a similar system to composers, where based on the content, the component could either require or not require close confirmation? That feels more light weight and practical.

One more thing: do you know if stacking modals is possible with `detail`? If so, I'd really like to consider having that: it's incredibly annoying when I have a modal, get an error, close debug, and then lose my original modal.

**Comment by davwheat** [comment]
How would we decide what "content" is worthy of requiring close confirmation?

I think that no matter what we should always have the option to override whatever logic is used to decide this.

Native dialogs have support for stacking, backdrops, and more! 

Take a look at the Chrome team's demos: https://googlechrome.github.io/dialog-polyfill/

The only downside is that we need a polyfill for full support in Chrome, but the Chrome team has made one of them anyway.

**Comment by askvortsov1** [comment]
> Native dialogs have support for stacking, backdrops, and more!

Beautiful!

> The only downside is that we need a polyfill for full support in Chrome, but the Chrome team has made one of them anyway. (For some reason Chrome Android supports it natively though...)

If it works, I'm fine with it :grin: Bundle size looks to be ~11kb, and I believe that should pay for itself when we scrap bootstrap?

> How would we decide what "content" is worthy of requiring close confirmation?

Look at how the composer does it: each composer body type can provide a callback that checks whether closing should be prevented: https://github.com/flarum/core/blob/a00c0ac6aa66e218b4f961fcd1d8c58fd6535126/js/src/forum/components/ComposerBody.js#L42-L42

**Comment by davwheat** [comment]
This PR.is deprecated in favour of using HTML5 dialog elements. Tracked in issue #<I can't find this easily on mobile>
